core/src/main/scala/cats/FunctorFilter.scala

```diff
@@ -0,0 +1,62 @@
+package cats
+
+import simulacrum.typeclass
+
+@typeclass trait FunctorFilter[F[_]] extends Functor[F] {
+
+  /**
+   * A combined [[map]] and [[filter]]. Filtering is handled via `Option`
+   * instead of `Boolean` such that the output type `B` can be different than
+   * the input type `A`.
+   *
+   * Example:
+   * {{{
+   * scala> import cats.implicits._
+   * scala> val m: Map[Int, String] = Map(1 -> "one", 3 -> "three")
+   * scala> val l: List[Int] = List(1, 2, 3, 4)
+   * scala> def asString(i: Int): Option[String] = m.get(i)
+   * scala> l.mapFilter(i => m.get(i))
+   * res0: List[String] = List(one, three)
+   * }}}
+   */
+  def mapFilter[A, B](fa: F[A])(f: A => Option[B]): F[B]
```

Comment:
There was some similar discussion a long time ago [here](https://github.com/typelevel/cats/issues/842#issuecomment-178583594). I wonder what the valid laws for this construct would be and whether you would run into issues where just working with `Foldable` would mean implementations would tend to be inefficient.

@johnynek were you suggesting something like this instead of `FunctorFilter` or `TraverseFilter`? I would tend to think that something like this should be in addition to these type classes. As you've mentioned you couldn't have an instance of this type class for `Option` or `OptionT`, which I think is a pretty big limitation, they are common types to call `filter` or `collect` on. I also feel like there is some benefit to knowing these methods can never make your structure grow (they can only reduce it or leave it the same); however this is just a fuzzy feeling and not something that I can confidently say is a useful property.